### DISCUSSION TITLE
Is it possible to receive message as string

### DISCUSSION THREAD

**Discussion by kefyru** [uNetworking/uWebSockets.js#217] [Q&A] [answered]
I'm using uWebSockets server on node.js
Client app connect to socket and send some JSON data in string,
but in server a receive ARRAY in message.
How can i get message as string?

**Comment by sirganya** [comment, marked as answer]
You can do it this way.
```
const buffer = Buffer.from(message);
const string = buffer.toString();
```

**Reply by xdev00** [reply]
cool. thank you so much for this.

**Comment by GrosSacASac** [comment]
or
const string = (new TextDecoder()).decode(new DataView(message))

**Comment by sirganya** [comment]
I tested both methods as performance is an issue for me. I wanted to avoid degrading the performance of uWS with inefficient built in methods. Using Bench.js to compare the two it seems TextDecoder loses out. I'd be glad for any input on the quality of the test, the results surprised me.

The results are

```
{
  node: '12.8.0',
  v8: '7.5.288.22-node.16',
  uv: '1.30.1',
  zlib: '1.2.11',
  brotli: '1.0.7',
  ares: '1.15.0',
  modules: '72',
  nghttp2: '1.39.1',
  napi: '4',
  llhttp: '1.1.4',
  http_parser: '2.8.0',
  openssl: '1.1.1c',
  cldr: '35.1',
  icu: '64.2',
  tz: '2019a',
  unicode: '12.1'
}
Scores: (bigger is better)

Buffer
Raw:
 > 5101.898101898102
 > 3376.6233766233768
 > 5088.911088911089
 > 4853.146853146854
Average (mean) 4605.144855144856

TextDecoder
Raw:
 > 163.67265469061877
 > 139.0268123138034
 > 150.24875621890547
 > 160.51844466600198
Average (mean) 153.3666669723324

Winner: Buffer
Compared with next highest (TextDecoder), it's:
96.67% faster
30.03 times as fast
1.48 order(s) of magnitude faster
A LOT FASTER
```

The code I used is 

```
const arrayBuffer = (function() {
    const buffer = new ArrayBuffer(4);
    const bufView = new Uint8Array(buffer);

    for (let i = 0; i < 4; i++) {
        bufView[i] = i + 1;
    }

    return buffer;
})();

exports.compare = {
    Buffer      : function() {
        const buffer = Buffer.from(arrayBuffer);
        const string = buffer.toString();
        return string;
    },
    TextDecoder : function() {
        const string = new TextDecoder().decode(new DataView(arrayBuffer));
        return string;
    }
};
require('bench').runMain();
```

**Reply by uasan** [reply]
>  it seems TextDecoder loses out. I'd be glad for any input on the quality of the test, the results surprised me.

Yes it is, they know and it hasn't been fixed yet.
https://github.com/nodejs/node/issues/39879

**Comment by ghost** [comment]
Yes you can do Buffer.from(message).toString() or any such alternative. uWS only deals with ArrayBuffer to/from JS since that representation maps 1:1 with the internals (there is no distinction between "string" and "binary" internally).

**Comment by fabiel-leon** [comment]
@alexhultman Could you put this in the documentation and samples ?
so people go straight to the doc and don't have to search the internet how to convert ArrayBuffer message to String,
I'm a bit of a novice at this excuse me if the answer is too simple

**Comment by ghost** [comment]
Yes it could make sense to update the texts on the docs

**Comment by zorro-fr24** [comment]
since this thread has been revived, I will point out that:

`String.fromCharCode.apply(null, message);`

is about 10-15% faster that `Buffer.from(message).toString();`

And that the slow dog TextDecoder, whilst still a slow dog would be 4 times faster than listed in the bench above if the class wasn't instantiated with every iteration ( and DataView can be dropped )

**Comment by ghost** [comment]
The fastest is of course to never make it a String, only dealing with binary protocols, but I understand JSON is everywhere so sadly this has to happen in most cases. We can add more info in the docs.

**Comment by zorro-fr24** [comment]
Indeed. 

String.fromCharCode is a little restrictive with character sets, but can be used to decode arraybuffer on the browser, which might allow the server to not have to deal with strings quite as much in that instance.
of course protobuf can also decode arraybuffers directly, unlike JSON.parse() which only takes strings or node buffer objects

**Comment by ghost** [comment]
Buffer.from(arrayBuffer).toJSON() should potentially be a fast path for ArrayBuffer -> JSON?

**Comment by uasan** [comment]
> Buffer.from(arrayBuffer).toJSON() should potentially be a fast path for ArrayBuffer -> JSON?

Yes, it’s default use.
```JS
JSON.parse(Buffer.from(arrayBuffer))
```
